File: src/test/java/com/winterhate/strategy/handlers/HandlersConfiguration.java
package com.winterhate.strategy.handlers;

import com.winterhate.strategy.api.StrategyDispatcher;
import com.winterhate.strategy.api.StrategyExecutor;
import com.winterhate.strategy.api.StrategySelector;
import com.winterhate.strategy.service.dispatcher.DefaultStrategyDispatcher;
import com.winterhate.strategy.service.executors.AsyncStrategyExecutor;
import com.winterhate.strategy.service.executors.DefaultStrategyExecutor;
import com.winterhate.strategy.service.executors.LoggingStrategyExecutor;
import com.winterhate.strategy.service.selectors.EnumStrategySelector;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.List;

@Configuration
public class HandlersConfiguration {

    @Bean
    public StrategySelector<HandlerDto, HandlerStrategy> handlerStrategySelector(List<HandlerStrategy> strategies) {
        return new EnumStrategySelector<>(strategies);
    }

    @Bean
    public StrategyExecutor<HandlerDto> handlerStrategyExecutor() {
        return new AsyncStrategyExecutor<>(new LoggingStrategyExecutor<>(new DefaultStrategyExecutor<>()));
    }

    @Bean
    public StrategyDispatcher<HandlerDto> handlerStrategyDispatcher(
            StrategySelector<HandlerDto, HandlerStrategy> handlerStrategySelector,
            StrategyExecutor<HandlerDto> handlerStrategyExecutor
    ) {
        return new DefaultStrategyDispatcher<>(handlerStrategySelector, handlerStrategyExecutor);
    }

}

File: src/test/kotlin/com/winterhate/strategy/handlers/HandlersConfiguration.kt
package com.winterhate.strategy.handlers

import com.winterhate.strategy.api.StrategyDispatcher
import com.winterhate.strategy.api.StrategyExecutor
import com.winterhate.strategy.api.StrategySelector
import com.winterhate.strategy.service.dispatcher.DefaultStrategyDispatcher
import com.winterhate.strategy.service.executors.AsyncStrategyExecutor
import com.winterhate.strategy.service.executors.DefaultStrategyExecutor
import com.winterhate.strategy.service.executors.LoggingStrategyExecutor
import com.winterhate.strategy.service.selectors.EnumStrategySelector
import org.slf4j.Logger
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class HandlersConfiguration {

    @Bean
    fun handlerStrategySelector(strategies: List<HandlerStrategy>): StrategySelector<HandlerDto, HandlerStrategy> =
        EnumStrategySelector(strategies)

    @Bean
    fun handlerStrategyExecutor(log: Logger): StrategyExecutor<HandlerDto> =
        AsyncStrategyExecutor(LoggingStrategyExecutor(DefaultStrategyExecutor(), log))

    @Bean
    fun handlerStrategyDispatcher(
        handlerStrategySelector: StrategySelector<HandlerDto, HandlerStrategy>,
        handlerStrategyExecutor: StrategyExecutor<HandlerDto>
    ): StrategyDispatcher<HandlerDto> =
        DefaultStrategyDispatcher(handlerStrategySelector, handlerStrategyExecutor)
}

